File: test/bench.js
// original file from:
// https://github.com/yyx990803/circular-json-es6

var dummy100 = require('./data.json');
var dummy50 = dummy100.slice(0, 50)
var dummy10 = dummy100.slice(0, 10)

function bench(method, dummy) {
  var t = Date.now(), i = 0
  while ((Date.now() - t) < 1000) {
    r = method(dummy)
    i++
  }
  return i
}

var r
var CircularJSON = require('circular-json')
var jsan = require('jsan')
var cj6 = require('circular-json-es6')
var flatted = require('../cjs')
var SC = require('@ungap/structured-clone/json')

function run(name, fn, dummy) {
  console.log(
    name + ' ' + (
      typeof dummy === 'string' ?
        dummy.length + ' chars' :
        (dummy.length || '') + ' objects'
    ) + ' parsed ' + (
      bench(fn, dummy) / 1000
    ).toFixed(2) + ' times per second'
  )
}

//*
console.log('-----------------------------------')
console.log('Object with ' + Object.keys(dummy100[0]).length + ' keys each')
console.log('-----------------------------------')
run('CircularJSON', CircularJSON.stringify, dummy100)
run('CircularJSON', CircularJSON.parse, r)
run('CircularJSON', CircularJSON.stringify, dummy50)
run('CircularJSON', CircularJSON.parse, r)
run('CircularJSON', CircularJSON.stringify, dummy10)
run('CircularJSON', CircularJSON.parse, r)
console.log('-----------------------------------')
run('circular-json-es6', cj6.stringify, dummy100)
run('circular-json-es6', cj6.parse, r)
run('circular-json-es6', cj6.stringify, dummy50)
run('circular-json-es6', cj6.parse, r)
run('circular-json-es6', cj6.stringify, dummy10)
run('circular-json-es6', cj6.parse, r)
console.log('-----------------------------------')
run('jsan', jsan.stringify, dummy100)
run('jsan', jsan.parse, r)
run('jsan', jsan.stringify, dummy50)
run('jsan', jsan.parse, r)
run('jsan', jsan.stringify, dummy10)
run('jsan', jsan.parse, r)
console.log('-----------------------------------')
run('flatted', flatted.stringify, dummy100)
run('flatted', flatted.parse, r)
run('flatted', flatted.stringify, dummy50)
run('flatted', flatted.parse, r)
run('flatted', flatted.stringify, dummy10)
run('flatted', flatted.parse, r)
console.log('-----------------------------------')
run('Structured Clone', SC.stringify, dummy100)
run('Structured Clone', SC.parse, r)
run('Structured Clone', SC.stringify, dummy50)
run('Structured Clone', SC.parse, r)
run('Structured Clone', SC.stringify, dummy10)
run('Structured Clone', SC.parse, r)
console.log('-----------------------------------')
console.log('50% same objects')
dummy100 = dummy50.concat(dummy50)
console.log('-----------------------------------')
run('CircularJSON', CircularJSON.stringify, dummy100)
run('CircularJSON', CircularJSON.parse, r)
run('circular-json-es6', cj6.stringify, dummy100)
run('circular-json-es6', cj6.parse, r)
run('jsan', jsan.stringify, dummy100)
run('jsan', jsan.parse, r)
run('flatted', flatted.stringify, dummy100)
run('flatted', flatted.parse, r)
run('Structured Clone', SC.stringify, dummy100)
run('Structured Clone', SC.parse, r)
console.log('-----------------------------------')
console.log('90% same objects')
dummy100 = [].concat(
  dummy10, dummy10, dummy10, dummy10, dummy10,
  dummy10, dummy10, dummy10, dummy10, dummy10
)
console.log('-----------------------------------')
run('CircularJSON', CircularJSON.stringify, dummy100)
run('CircularJSON', CircularJSON.parse, r)
run('circular-json-es6', cj6.stringify, dummy100)
run('circular-json-es6', cj6.parse, r)
run('jsan', jsan.stringify, dummy100)
run('jsan', jsan.parse, r)
run('flatted', flatted.stringify, dummy100)
run('flatted', flatted.parse, r)
run('Structured Clone', SC.stringify, dummy100)
run('Structured Clone', SC.parse, r)
console.log('-----------------------------------')
console.log('with circular')
function makeCircularObject () {
  var a = {}
  a.b = a
  return a
}
dummy100 = []
for (var i = 0; i < 100; i++) {
  dummy100.push(makeCircularObject())
}
console.log('-----------------------------------')
run('CircularJSON', CircularJSON.stringify, dummy100)
run('CircularJSON', CircularJSON.parse, r)
run('circular-json-es6', cj6.stringify, dummy100)
run('circular-json-es6', cj6.parse, r)
run('jsan', jsan.stringify, dummy100)
run('jsan', jsan.parse, r)
run('flatted', flatted.stringify, dummy100)
run('flatted', flatted.parse, r)
run('Structured Clone', SC.stringify, dummy100)
run('Structured Clone', SC.parse, r)
console.log('-----------------------------------')
console.log('with circular 90% same')
function makeCircularObject () {
  var a = {}
  a.b = a
  return a
}
dummy10 = []
for (var i = 0; i < 10; i++) {
  dummy10.push(makeCircularObject())
}
dummy100 = [].concat(
  dummy10, dummy10, dummy10, dummy10, dummy10,
  dummy10, dummy10, dummy10, dummy10, dummy10
)
console.log('-----------------------------------')
run('CircularJSON', CircularJSON.stringify, dummy100)
run('CircularJSON', CircularJSON.parse, r)
run('circular-json-es6', cj6.stringify, dummy100)
run('circular-json-es6', cj6.parse, r)
run('jsan', jsan.stringify, dummy100)
run('jsan', jsan.parse, r)
run('flatted', flatted.stringify, dummy100)
run('flatted', flatted.parse, r)
run('Structured Clone', SC.stringify, dummy100)
run('Structured Clone', SC.parse, r)
// */
console.log('-----------------------------------')
console.log('Big real-world circular data')
var cirular = CircularJSON.parse(
              require('fs').readFileSync(
                require('path').join(__dirname, './circular.txt')
              ).toString());
console.log('-----------------------------------')
run('CircularJSON', CircularJSON.stringify, cirular)
run('CircularJSON', CircularJSON.parse, r)
run('circular-json-es6', cj6.stringify, cirular)
run('circular-json-es6', cj6.parse, r)
run('jsan', jsan.stringify, cirular)
run('jsan', jsan.parse, r)
run('flatted', flatted.stringify, cirular)
run('flatted', flatted.parse, r)
run('Structured Clone', SC.stringify, cirular)
run('Structured Clone', SC.parse, r)

File: rust/examples/bench.rs
//! Flatted bench mirroring `test/bench.js` timing (1s wall-clock ops/sec).
//!
//! ```bash
//! cargo run --example bench --release
//! cargo run --example bench --release -- --json
//! ```

use std::cell::RefCell;
use std::env;
use std::fs;
use std::path::{Path, PathBuf};
use std::rc::Rc;
use std::time::{Duration, Instant};

use flatted::{parse_simple, stringify_simple, Value};
use serde_json::Value as JsonValue;

const WINDOW: Duration = Duration::from_secs(1);

fn repo_test_dir() -> PathBuf {
    // examples/ -> rust/ -> repo root -> test/
    Path::new(env!("CARGO_MANIFEST_DIR"))
        .join("..")
        .join("test")
}

fn bench_stringify(value: &Value) -> f64 {
    let start = Instant::now();
    let mut i = 0u64;
    let mut last = String::new();
    while start.elapsed() < WINDOW {
        last = stringify_simple(value).expect("stringify");
        i += 1;
    }
    // keep last result live so the optimizer cannot drop the work
    std::hint::black_box(&last);
    i as f64 / WINDOW.as_secs_f64()
}

fn bench_parse(text: &str) -> f64 {
    let start = Instant::now();
    let mut i = 0u64;
    let mut last = Value::Null;
    while start.elapsed() < WINDOW {
        last = parse_simple(text).expect("parse");
        i += 1;
    }
    std::hint::black_box(&last);
    i as f64 / WINDOW.as_secs_f64()
}

fn load_data_json() -> Value {
    let path = repo_test_dir().join("data.json");
    let raw = fs::read_to_string(&path).unwrap_or_else(|e| panic!("read {}: {e}", path.display()));
    let json: JsonValue = serde_json::from_str(&raw).expect("data.json");
    Value::from_json(&json)
}

fn slice_array(value: &Value, len: usize) -> Value {
    let items: Vec<Value> = value
        .as_array()
        .expect("array")
        .borrow()
        .iter()
        .take(len)
        .cloned()
        .collect();
    Value::Array(Rc::new(RefCell::new(items)))
}

fn concat_shared(parts: &[&Value]) -> Value {
    let mut out = Vec::new();
    for part in parts {
        out.extend(part.as_array().expect("array").borrow().iter().cloned());
    }
    Value::Array(Rc::new(RefCell::new(out)))
}

fn make_circular_object() -> Value {
    let o = Value::empty_object();
    o.as_object()
        .unwrap()
        .borrow_mut()
        .insert("b".into(), o.clone());
    o
}

fn circular_list(n: usize) -> Value {
    let items: Vec<Value> = (0..n).map(|_| make_circular_object()).collect();
    Value::Array(Rc::new(RefCell::new(items)))
}

struct ResultRow {
    case: &'static str,
    op: &'static str,
    label: String,
    ops: f64,
}

fn emit(row: &ResultRow, json: bool) {
    if json {
        println!(
            r#"{{"engine":"rust","case":"{}","op":"{}","label":"{}","ops":{:.2}}}"#,
            row.case,
            row.op,
            row.label.replace('"', "\\\""),
            row.ops
        );
    } else {
        println!(
            "rust / flatted {} {} parsed {:.2} times per second",
            row.op, row.label, row.ops
        );
    }
}

fn pair(case: &'static str, label: String, value: &Value) -> (ResultRow, ResultRow) {
    let stringify_ops = bench_stringify(value);
    let text = stringify_simple(value).unwrap();
    let parse_ops = bench_parse(&text);
    (
        ResultRow {
            case,
            op: "stringify",
            label: label.clone(),
            ops: stringify_ops,
        },
        ResultRow {
            case,
            op: "parse",
            label,
            ops: parse_ops,
        },
    )
}

fn main() {
    let json = env::args().any(|a| a == "--json");
    let data = load_data_json();
    let keys = data.as_array().unwrap().borrow()[0]
        .as_object()
        .unwrap()
        .borrow()
        .len();

    if !json {
        println!("-----------------------------------");
        println!("Object with {keys} keys each");
        println!("-----------------------------------");
    }

    let dummy100 = data.clone();
    let dummy50 = slice_array(&data, 50);
    let dummy10 = slice_array(&data, 10);

    for (case, label, value) in [
        ("objects_100", "100 objects".to_string(), &dummy100),
        ("objects_50", "50 objects".to_string(), &dummy50),
        ("objects_10", "10 objects".to_string(), &dummy10),
    ] {
        let (s, p) = pair(case, label, value);
        emit(&s, json);
        emit(&p, json);
    }

    if !json {
        println!("-----------------------------------");
        println!("50% same objects");
        println!("-----------------------------------");
    }
    let shared50 = concat_shared(&[&dummy50, &dummy50]);
    let (s, p) = pair("shared_50", "100 objects".into(), &shared50);
    emit(&s, json);
    emit(&p, json);

    if !json {
        println!("-----------------------------------");
        println!("90% same objects");
        println!("-----------------------------------");
    }
    let shared90 = concat_shared(&[
        &dummy10, &dummy10, &dummy10, &dummy10, &dummy10, &dummy10, &dummy10, &dummy10, &dummy10,
        &dummy10,
    ]);
    let (s, p) = pair("shared_90", "100 objects".into(), &shared90);
    emit(&s, json);
    emit(&p, json);

    if !json {
        println!("-----------------------------------");
        println!("with circular");
        println!("-----------------------------------");
    }
    let circ = circular_list(100);
    let (s, p) = pair("circular_100", "100 objects".into(), &circ);
    emit(&s, json);
    emit(&p, json);

    if !json {
        println!("-----------------------------------");
        println!("with circular 90% same");
        println!("-----------------------------------");
    }
    let circ10 = circular_list(10);
    let circ90 = concat_shared(&[
        &circ10, &circ10, &circ10, &circ10, &circ10, &circ10, &circ10, &circ10, &circ10, &circ10,
    ]);
    let (s, p) = pair("circular_shared_90", "100 objects".into(), &circ90);
    emit(&s, json);
    emit(&p, json);

    // Big real-world: prefer preconverted flatted payload (from compare harness).
    let flatted_path = repo_test_dir().join(".bench-circular.flatted.json");
    if flatted_path.is_file() {
        if !json {
            println!("-----------------------------------");
            println!("Big real-world circular data");
            println!("-----------------------------------");
        }
        let text = fs::read_to_string(&flatted_path).expect("circular flatted");
        let value = parse_simple(&text).expect("parse circular flatted");
        let label = format!("{} chars", text.len());
        let stringify_ops = bench_stringify(&value);
        let parse_ops = bench_parse(&text);
        emit(
            &ResultRow {
                case: "big_circular",
                op: "stringify",
                label: label.clone(),
                ops: stringify_ops,
            },
            json,
        );
        emit(
            &ResultRow {
                case: "big_circular",
                op: "parse",
                label,
                ops: parse_ops,
            },
            json,
        );
    } else if !json {
        println!("-----------------------------------");
        println!(
            "Big real-world circular data (skipped: missing {})",
            flatted_path.display()
        );
        println!("Run: node test/bench-js-vs-rs.mjs   # prepares the fixture");
        println!("-----------------------------------");
    }
}
